spec.md

```diff
@@ -0,0 +1,276 @@
+
+### POST `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+### PUT `/ipns/<key>[/<path>]`
+
+The same as POST `/ipns/<key>[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### PUT `/ipns/localhost[/<path>]`
+
+The same as POST `/ipns/localhost[/<path>]`, but where the response body is the file to be uploaded, not an IPFS path. Multiple files can be uploaded using form data, in which case files will go inside the directory in the path.
+
+### DELETE `/ipns/<key>[/<path>]`
+
+Remove content behind an IPNS name. The new content is pinned, and the previous content is unpinned.
+
+#### Response
+A 301 Temporary Redirect to `ipns://<key>[/<path>]`. The path in the redirect is the directory containing the deletion path. So if the path in the URL was `/ipns/<key>/path/to/file.ext`, the redirect will be to `ipns://<key>/path/to`.
+
+The header `X-IPNS-Path` is set to `/ipns/<key>[/<path>]`, with the same path logic as above.
+
+### DELETE `/ipns/localhost[/<path>]`
+
+Behaves the same as `/ipns/<key>[/<path>]`.
+
+#### Query params
+- `key=some_name`: **Required**. This specifies the key you want by a name instead of by public key. If a key with this name doesn't exist, it will be created.
+
+#### Response
+The same as `/ipns/<key>[/<path>]`. Note this allows for extraction of the actual public key from future use and link sharing.
+
+## Pubsub
+
+An experimental pubsub API is supported to give users direct access to sending and receiving pubsub messages.
+
+### GET `/pubsub/<topic>`
+
+#### Query params
+- `format`: can be `json`, `utf8`, `base64`, or not specified, which defaults to `base64`. This controls the decoding of pubsub message bytes
+
+#### Response
+
+The response is an [event stream](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format), so pubsub messages for that topic will be sent to the client as long as the connection remains open.
+
+Pubsub messages are encoded as events like this:
+
+```
+id: FuhB2orEIwE=
+data: {"from":"QmfULfLaRtF7Kg3ShjFdiYL97P7eEcyL4PcvQLhkzCPCeG","data":"aGVsbG8=","topics":["test"]}
+```
+
+The "message identifier" of the pubsub message is base64-encoded to form the `id:`. `data:` is always encoded as JSON, with the other three things known about a pubsub message.
+
+In the JSON: `from` is the prettified ID of the node who sent the message, `data` is the message data, and `topics` is an array of topics this message was sent to.
```

Comment:
nit: are you limiting topic names so they can't produce malformed JSON? if not, it shouls also be  wrapped in multibase 

See prior art in https://github.com/ipfs/go-ipfs/issues/7939#issuecomment-833812863 and https://github.com/ipfs/go-ipfs-cmds/issues/224

Reply:
The topic names are encoded as JSON strings. That means any invalid values will be escaped and handled fine, no multibase encoding needed. A newline becomes `\n`, `"` becomes `\"`, `[` stays as `[`.